internal/app/starter/master_linux.go

```diff
@@ -11,129 +11,131 @@ import (
+
+			if err != io.EOF {
+				fatalChan <- fmt.Errorf("error while reading master socket data: %s", err)
 				return
 			}
-		}
-		// wait container process execution, EOF means container process
-		// was executed and master socket was closed by stage 2. If data
-		// byte sent is equal to 'f', it means an error occurred in
-		// StartProcess, just return by waiting error and process status
-		_, err = conn.Read(data)
-		if (err != nil && err != io.EOF) || data[0] == 'f' {
+			// EOF means something goes wrong in stage 2, don't send error via
+			// fatalChan, error will be reported by stage 2 and the process
+			// status will be set accordingly via MonitorContainer method below
+			sylog.Debugf("stage 2 process was interrupted, waiting status")
+			return
+		} else if data[0] == 'f' {
+			// StartProcess reported an error in stage 2, don't send error via
+			// fatalChan, error will be reported by stage 2 and the process
+			// status will be set accordingly via MonitorContainer method below
 			sylog.Debugf("stage 2 process reported an error, waiting status")
 			return
 		}
-
-		err = e.PostStartProcess(containerPid)
-		if err != nil {
-			fatalChan <- fmt.Errorf("post start process failed: %s", err)
+		if err := obj.PreStartProcess(containerPid, conn, fatalChan); err != nil {
+			fatalChan <- fmt.Errorf("pre start process failed: %s", err)
 			return
 		}
-		if isInstance && os.Getppid() == ppid {
-			// sleep a bit to see if child exit
-			time.Sleep(100 * time.Millisecond)
-			syscall.Kill(ppid, syscall.SIGUSR1)
-		}
-	}()
+	}
+	// wait container process execution, EOF means container process
+	// was executed and master socket was closed by stage 2. If data
+	// byte sent is equal to 'f', it means an error occurred in
+	// StartProcess, just return by waiting error and process status
+	_, err = conn.Read(data)
+	if (err != nil && err != io.EOF) || data[0] == 'f' {
+		sylog.Debugf("stage 2 process reported an error, waiting status")
+		return
+	}
+
+	err = e.PostStartProcess(containerPid)
+	if err != nil {
+		fatalChan <- fmt.Errorf("post start process failed: %s", err)
+		return
+	}
+}
+
+// Master initializes a runtime engine and runs it
+func Master(rpcSocket, masterSocket int, containerPid int, e *engine.Engine) {
+	var status syscall.WaitStatus
+	fatalChan := make(chan error, 1)
+
+	// we could receive signal from child with CreateContainer call so we
+	// set the signal handler earlier to queue signals until MonitorContainer
+	// is called to handle them
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals)
+
+	go createContainer(rpcSocket, containerPid, e, fatalChan)
+
+	go startContainer(masterSocket, containerPid, e, fatalChan)
 
 	go func() {
 		var err error
 		status, err = e.MonitorContainer(containerPid, signals)
 		fatalChan <- err
 	}()
 
-	fatal = <-fatalChan
+	fatal := <-fatalChan
 
-	runtime.LockOSThread()
 	if err := e.CleanupContainer(fatal, status); err != nil {
```

Comment:
Same question, why removing `runtime.LockOSThread`?

Reply:
@sashayakovtseva I do not understand your request here, are you asking to add comments?